File: leetcode/max_subarray_len2.py
#!/usr/bin/env python3

from collections import defaultdict
import sys
import unittest

# number: 325
# section: meta
# difficulty: medium
# tags: array, hash table, prefix sum, meta

# constraints
# 1 <= nums.length <= 10^5
# -10^4 <= nums[i] <= 10^4
# -10^4 <= k <= 10^4

# solution: prefix sum + defaultdict
# complexity
# run-time: O(n)
# space: O(n)
def max_subarray_len2(nums, k) -> int:
    if not nums:
        return 0

    # calculate prefix sum
    # TODO: don't build full prefix sum, use a variable?
    prefix = [nums[0]]

    for i in range(1, len(nums)):
        prefix.append(prefix[-1]+nums[i])

    #print(f"prefix:{prefix}")

    d = {}
    ans = 0
       
    for i in range(len(prefix)):
        # do check if sum so far is equal to k
        if prefix[i] == k:
            # add 1 because we want to count the first index
            ans = i + 1

        # like two sum
        sum = prefix[i]-k
        if sum in d:
            ans = max(ans, i-d[sum])

        # don't prebuild
        # don't overwrite the first index of prefix sum
        # because we want to keep the first index
        # (to maximize len)
        if prefix[i] not in d:
            d[prefix[i]] = i
            
    return ans

class TestMaxSubarrayLen2(unittest.TestCase):
    def test_empty(self):
        nums = []
        k = 0
        expected = 0
        self.assertEqual(max_subarray_len2(nums, k), expected)

    def test1(self):
        nums = [1,-1,5,-2,3]
        k = 3
        expected = 4
        self.assertEqual(max_subarray_len2(nums, k), expected)

    def test2(self):
        nums = [-2,-1,2,1]
        k = 1
        expected = 2
        self.assertEqual(max_subarray_len2(nums, k), expected)

if __name__ == "__main__":
    sys.exit(unittest.main())

File: leetcode/max_subarray_len2.cpp
#include <algorithm>
#include <vector>
#include <unordered_map>

// number: 325
// title: Maximum Size Subarray Sum Equals k
// url: https://leetcode.com/problems/maximum-size-subarray-sum-equals-k/
// section: meta
// difficulty: medium
// tags: array, hash table, prefix sum, meta

// constraints
// 1 <= nums.length <= 10^5
// -10^4 <= nums[i] <= 10^4
// -10^4 <= k <= 10^4

// solution: prefix sum + unodered_map
// complexity
// run-time: O(n)
// space: O(n)
int maxSubArrayLen2(std::vector<int>& nums, int k)
{
    if (nums.empty()) {
        return 0;
    }

    std::vector<long long> prefix{nums[0]};

    for (size_t i = 1; i != nums.size(); ++i) {
        prefix.push_back(prefix.back()+nums[i]);
    }

    // don't use map
    // don't overflow
    std::unordered_map<int, long long> prefix2index;
    long long ans = 0;

    for (long long i = 0; i != prefix.size(); ++i) {
        if (prefix[i] == k) {
            ans = i + 1;
        }

        long long sum = prefix[i] - k;
        auto it = prefix2index.find(sum);
        if (it != prefix2index.end()) {
            ans = std::max(ans, i-prefix2index[sum]);
        }

        // don't prebuild, because we need to find the first index of prefix[i]
        it = prefix2index.find(prefix[i]);
        if (it == prefix2index.end()) {
            prefix2index[prefix[i]] = i;
        }
    }

    return ans;
}

// TOOD: add test cases